### DISCUSSION TITLE
Properly starting async connection with SQLAlchemy Core in FastAPI (ASGI)

### DISCUSSION THREAD

**Discussion by jocho-here** [sqlalchemy/sqlalchemy#7473] [Usage Questions]
I'm trying to build an async backend API server using FastAPI, Postgres, and SQLAlchemy Core.

When I was testing my API server without the frontend, it ran perfectly, but as soon as the frontend started to hit the backend concurrently, I started seeing the following error:

```
ERROR:    Exception in ASGI application
...
sqlalchemy.exc.ResourceClosedError: This Connection is closed
```

Here's my setup:

1. service/utils/db.py
```
from fastapi import Request
from sqlalchemy.ext.asyncio import create_async_engine
from service.utils.config import DB_DSN
...
def get_async_engine(echo: bool = True, future: bool = Ture):
    global ENGINE

    if not ENGINE:
        ENGINE = create_async_engine(
            str(DB_DSN), echo=echo, future=future
        )

    return ENGINE

async def get_async_conn():
    engine = get_async_engine()

    async with engine.begine() as conn:
        Request.db_conn = conn
        yield
```

2. service/api/main.py
```
from fastapi import Depends, FastAPI

from service.utils.db import get_async_conn

def get_app():
    app = FastAPI(
        title="app", dependencies=[Depends(get_async_conn)]
    )
    ...

    return app
```

3. Example endpoint
```
from fastapi import APIRouter

router = APIRouter(...)

@router.get("")
async def get_items():
    query = await Request.db_conn.execute(...)  # some SQL query
    items = query.fetchall()
```

Any suggestion on how to **properly** manage async engine & connection with SQLAlchemy Core?

**Comment by CaselIT** [comment]
Hi,

Not familiar with FastApi, but won't this `Request.db_conn = conn` set the connection grobally? you can't do that when doing things concurrently.

**Reply by jocho-here** [reply]
These [global dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/) are called on every endpoint hit, so the connection should be renewed every time an endpoint is hit.

**Reply by CaselIT** [reply]
Sure, but the issue is not how many times the function is called, since I assumed it was called each time.
The issue seems to be that request 1 will set its connection as a global, request 2 will the run e replace the first connection with another one, ect 

As mentioned above, I you should not use global objects for the connections that are local to a single request.

**Reply by CaselIT** [reply]
A recent discussion also linked this tutorial for fastapi https://github.com/mdhishaamakhtar/fastapi-sqlalchemy-postgres-template/ that locally uses a session for each request

**Reply by jocho-here** [reply]
This was a good suggestion. I started creating engine every time there was a new request by: 
```
def get_async_engine(echo: bool = True, future: bool = Ture):
    return create_async_engine(str(DB_DSN), echo=echo, future=future)
```

This, however, didn't fix the issue. With this, I still saw the problem of the connection closed error, with 500 on the frontend network console.

I am trying to resolve this by spinning multiple processes of API servers using `gunicorn` and `uvicorn` even in local.

I don't think this will always work. Let me know if you are using SQLAlchemy's async engine raw and made it working with your ASGI Framework!

**Reply by CaselIT** [reply]
> I started creating engine every time there was a new request by:

this is really not how sqlalchemy is supposed to be used. 

I would suggest asking for help in the api server you are using

**Reply by jocho-here** [reply]
Ah, I think I wasn't fully understanding how threads, requests, and db connections are entangled.

I've printed out ids of every request's thread, the connection I create, and `Request.db_conn` I set up.

It turns out, all requests are being handled by a single thread (I'm just running FastAPI with single process, uvicorn). When there are concurrent frontend requests, `Request.db_conn = conn` gets called multiple times before ending the request's session.
This replaced the thread's `Request.db_conn` multiple times and this made whatever DB processing to hit the latest assigned DB connection at `Request.db_conn`. Some of the DB queries ran were ran on a wrong connection. And also, `engine.dispose` cause this Connection closed issue, since `Request.db_conn` may have the earlier engine's connection, which has already been closed.

This is what @CaselIT already said, I think. I just didn't understand without knowing how they are set.

Do you guys know if there's a way to assign/ set DB connection to that specific request, without having Request passed around like how it's explained in the documentation [here](https://fastapi.tiangolo.com/advanced/using-request-directly/?h=request)?

It looks like using encode's [databases](https://github.com/encode/databases) can help me achieve this, but not sure if I want to introduce whole lot of change just yet.

**Reply by CaselIT** [reply]
You could potentially use context vars. https://docs.python.org/3/library/contextvars.html

But in general there are many disadvantages to using global magic variables like you are trying to, compared to explicitly passing around the objects you need.
In any case if you think it's best for your use case, but could also look into flask server that uses that model

**Comment by gordthompson** [comment]
I don't know if this is part of your issue, but you are invoking the dependency to get the async session/connection at the app level while most examples I've seen invoke the dependency at the route (path operation) level, e.g., 

https://github.com/gordthompson/fastapi-tutorial-cockroachdb-async/blob/66f52378cae6fa5e8080d144e157ed0ac2142496/sql_app/main.py#L21

**Reply by jocho-here** [reply]
Hey, I've seen many examples, and I was trying to come up with a way to be able to use DB connection anywhere without passing the connection around.
